File: lib/java/src/main/java/io/github/brenbar/uapi/internal/DecodeKeys.java
package io.github.brenbar.uapi.internal;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class DecodeKeys {
    static Object decodeKeys(Object given, BinaryEncoding binaryEncoder) {
        if (given instanceof Map<?, ?> m) {
            final var newMap = new HashMap<String, Object>();

            for (final var e : m.entrySet()) {
                final String key;
                if (e.getKey() instanceof final String s) {
                    key = s;
                } else {
                    key = (String) binaryEncoder.decodeMap.get(e.getKey());

                    if (key == null) {
                        throw new BinaryEncodingMissing(key);
                    }
                }
                final var encodedValue = decodeKeys(e.getValue(), binaryEncoder);

                newMap.put(key, encodedValue);
            }

            return newMap;
        } else if (given instanceof final List<?> l) {
            return l.stream().map(e -> decodeKeys(e, binaryEncoder)).toList();
        } else {
            return given;
        }
    }
}

File: tool/uapicodegen/uapicodegen/uapi/internal/binary/DecodeKeys.py
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from ...internal.binary.BinaryEncoding import BinaryEncoding


def decode_keys(given: object, binary_encoder: 'BinaryEncoding') -> object:
    from ...internal.binary.BinaryEncodingMissing import BinaryEncodingMissing

    if isinstance(given, dict):
        new_dict: dict[str, object] = {}

        for key, value in given.items():
            if isinstance(key, str):
                new_key = key
            else:
                possible_new_key = binary_encoder.decode_map.get(key)

                if possible_new_key is None:
                    raise BinaryEncodingMissing(key)

                new_key = possible_new_key

            encoded_value = decode_keys(value, binary_encoder)
            new_dict[new_key] = encoded_value

        return new_dict
    elif isinstance(given, list):
        return [decode_keys(item, binary_encoder) for item in given]
    else:
        return given
